### PR TITLE
Initial draft stol apt repostory setup for carma-geodetic-lib

### PR DESCRIPTION
**Pull request by paulbourelly999**
# PR Details
## Description

> [!IMPORTANT]\
> Waiting on CPM Support PR (https://github.com/usdot-fhwa-stol/carma-builds/pull/24), Packing.cmake fix PR (https://github.com/usdot-fhwa-stol/carma-builds/pull/26) and Fixes to the Testing.cmake fix PR (https://github.com/usdot-fhwa-stol/carma-builds/pull/27)

This Pull Request is for moving some of the logic in the[ CARMA Platform cooperative perception package](https://github.com/usdot-fhwa-stol/carma-platform/tree/develop/carma_cooperative_perception) to an independent library that is ROS independent and integrated with our STOL apt debian package repository CI process (See [carma-builds](https://github.com/usdot-fhwa-stol/carma-builds) for more information). The logic that is being included specifically in this library/repository is logic used to convert WSG84 coordinates to UTM coordinates and offsets. This will currently be used for processing SDSMs on the vehicle and infrastructure sides on CARMA Platform and CARMA Streets respectively.

## Related Jira Key
[CDAR-352
](https://usdot-carma.atlassian.net/browse/CDAR-352)

## Motivation and Context
Move logic into independent library that can be included in both CARMA Platform and CARMA Streets for VRU UC1.

## How Has This Been Tested?
Locally built in devcontainer.

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+				"EditorConfig.EditorConfig",
+				"twxs.cmake",
+				"ms-vscode.cmake-tools"
+			]
+		}
+	}
+
+	// Uncomment to connect as an existing user other than the container default. More info: https://aka.ms/dev-containers-non-root.
+	// "remoteUser": "devcontainer"
+}
```
Comment (adamlm) [inline .devcontainer/devcontainer.json]: Let's remove the boilerplate comments to avoid cluttering the file.

File: .editorconfig [hunk 2]
```diff
@@ -0,0 +1,15 @@
+# EditorConfig <http://EditorConfig.org>
```
Comment (adamlm) [inline .editorconfig:1]: Is there a particular reason we are using EditorConfig? The VS Code extensions appears abandoned, there hasn't been an update in over 2 years. Some of these configurations are better left to other tools such as clang-format for C++ files.

I suggest we use Prettier, which provides the same general functionality as EditorConfig but is much more active.

File: .vscode/launch.json [hunk 3]
```diff
@@ -0,0 +1,28 @@
+{
```
Comment (adamlm) [inline .vscode/launch.json]: I don't think we need this configuration. The VS Code CMake extension provides an easy way to launch debugging. Am I missing something here?

File: .vscode/tasks.json [hunk 4]
```diff
@@ -0,0 +1,26 @@
+{
```
Comment (adamlm) [inline .vscode/tasks.json:1]: Is this task used for anything? It looks like the `build.sh` file relies on something in `/opt/carma`, but it's unclear why it needs to be  VS COde task.

File: CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,55 @@
+cmake_minimum_required(VERSION 3.10.2)
+
+# where to find our CMake modules
+set(CMAKE_MODULE_PATH "$ENV{CARMA_OPT_DIR}/cmake")
```
Comment (adamlm) [inline CMakeLists.txt]: We should not be overwriting `CMAKE_MODULE_PATH`. We should append to it instead.

```cmake
list(APPEND CMAKE_MODULE_PATH $ENV{CAMRA_OPT_DIR}/cmake)
```
We should also check to see if its defined first:

```cmake
if(DEFINED ENV{CARMA_OPT_DIR})
  list(APPEND CMAKE_MODULE_PATH $ENV{CAMRA_OPT_DIR}/cmake)
endif()
```

File: CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+# Needed so CMake can find the vendored PROJ4 module file. Th FindPROJ4.cmake
+# module file can be removed if we upgrade to a more recent PROJ version. See
+# https://github.com/usdot-fhwa-stol/carma-platform/issues/2139 for updates.
+list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/cmake)
+
+project(carma-geodetic
+    DESCRIPTION "CARMA geodetic library"
+    HOMEPAGE_URL https://github.com/usdot-fhwa-stol/carma-geodetic-lib
+	VERSION 0.0.1
```
Comment (adamlm) [inline CMakeLists.txt]: Semantic versioning guidelines recommend 0.1.0 for initial development releases. [Link](https://semver.org/#how-should-i-deal-with-revisions-in-the-0yz-initial-development-phase)

File: CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,55 @@
+cmake_minimum_required(VERSION 3.10.2)
```
Comment (adamlm) [inline CMakeLists.txt:1]: CMake uses 2-space indents.

File: CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+project(carma-geodetic
+    DESCRIPTION "CARMA geodetic library"
+    HOMEPAGE_URL https://github.com/usdot-fhwa-stol/carma-geodetic-lib
+	VERSION 0.0.1
+    LANGUAGES CXX
+)
+
+find_package(Threads)
+
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pthread")
```
Comment (adamlm) [inline CMakeLists.txt]: I don't see anywhere in the code that requires us to link specifically with `pthread`. This should be removed.

File: CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+include(cmake/dependencies.cmake)
+
+add_library(${PROJECT_NAME} SHARED)
```
Comment (adamlm) [inline CMakeLists.txt:20]: Coupling the project name to the library name is a frowned-upon practice. It also makes reading the CMake configurations more difficult. This and all references to ${PROJECT_NAME} should be replaced with the intended target name, even if it matches the project's name.

File: CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+ ${PROJ4_LIBRARIES}  # Note: Newer versions of PROJ use PROJ::proj
+  GSL  # Note: Newer versions of GSL use Microsoft.GSL::GSL
+  units::units
+)
+target_include_directories(${PROJECT_NAME} 
+    PUBLIC
+        $<INSTALL_INTERFACE:include>
+        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
+    PRIVATE
+        ${CMAKE_CURRENT_SOURCE_DIR}/src
```
Comment (adamlm) [inline CMakeLists.txt]: We do not need to add this. There are no private headers in `src/`.

File: CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+add_library(${PROJECT_NAME} SHARED)
+target_link_libraries( ${PROJECT_NAME}
+ ${PROJ4_LIBRARIES}  # Note: Newer versions of PROJ use PROJ::proj
+  GSL  # Note: Newer versions of GSL use Microsoft.GSL::GSL
+  units::units
+)
+target_include_directories(${PROJECT_NAME} 
+    PUBLIC
+        $<INSTALL_INTERFACE:include>
+        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
```
Comment (adamlm) [inline CMakeLists.txt]: I think changing this to ${PROJECT_SOURCE_DIR} would be more appropriate.

File: CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+add_library(${PROJECT_NAME} SHARED)
+target_link_libraries( ${PROJECT_NAME}
+ ${PROJ4_LIBRARIES}  # Note: Newer versions of PROJ use PROJ::proj
+  GSL  # Note: Newer versions of GSL use Microsoft.GSL::GSL
+  units::units
+)
+target_include_directories(${PROJECT_NAME} 
+    PUBLIC
+        $<INSTALL_INTERFACE:include>
```
Comment (adamlm) [inline CMakeLists.txt]: This can be removed. This is an older convention, and the functionality has been incorporated into the `install(TARGET)` command.

File: CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+find_package(Threads)
+
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pthread")
+
+include(cmake/dependencies.cmake)
+
+add_library(${PROJECT_NAME} SHARED)
+target_link_libraries( ${PROJECT_NAME}
+ ${PROJ4_LIBRARIES}  # Note: Newer versions of PROJ use PROJ::proj
+  GSL  # Note: Newer versions of GSL use Microsoft.GSL::GSL
```
Comment (adamlm) [inline CMakeLists.txt]: Remove the comment since it is no longer accurate.

Also, we should be linking against `Microsoft.GSL::GSL` if this version supports it.

File: CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+        $<INSTALL_INTERFACE:include>
+        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
+    PRIVATE
+        ${CMAKE_CURRENT_SOURCE_DIR}/src
+)
+target_link_libraries( ${PROJECT_NAME}
+ ${PROJ4_LIBRARIES}  # Note: Newer versions of PROJ use PROJ::proj
+  GSL  # Note: Newer versions of GSL use Microsoft.GSL::GSL
+  units::units
+)
```
Comment (adamlm) [inline CMakeLists.txt]: Remove this because it is a duplicate.

File: CMakeLists.txt [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+set_target_properties(${PROJECT_NAME} PROPERTIES
+    SOVERSION 1
+)
```
Comment (adamlm) [inline CMakeLists.txt]: I don't think this is needed.

File: README.md [hunk 16]
```diff
@@ -1 +1,100 @@
...
+| .devcontainer/devcontainer.json       | Docker Dev Environment definition file.                                                                              |  |  |  |
+| .editorconfig                         | Some standard, robust VSCode editor configurations.                                                                  |  |  |  |
+| .github/workflows/build.yml           | GitHub CI build definition.                                                                                          |  |  |  |
+| .github/workflows/sonar-scanner.yml   | TODO                                                                                                                 |  |  |  |
+| .gitignore                            | Some helpful Git Ignore rules that should apply to Cmake-based C++ projects developed with most modern code editors. |  |  |  |
+| install_dependencies.sh               | Docker Dev Environment internal script. TODO move this file?                                                         |  |  |  |
+| .vscode/tasks.json                    | Build task definition utilizing Docker Dev Environment.                                                              |  |  |  |
```
Comment (adamlm) [inline README.md]: Remove unused columns

File: build.sh [hunk 17]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+
+set -e
+
+${CARMA_OPT_DIR}/scripts/build_script.sh -p $@
```
Comment (adamlm) [inline build.sh]: Is this file needed? What is it for?

File: install_dependencies.sh [hunk 18]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
```
Comment (adamlm) [inline install_dependencies.sh]: If this file needs bash capabilities, it should be renamed to `install_dependencies.bash`. Otherwise, the shebang should be changed to `#!/bin/sh`.

File: install_dependencies.sh [hunk 19]
```diff
@@ -0,0 +1,10 @@
...
+
+set -ex
+# Direct Package Dependencies
+DEPENDENCIES="libproj-dev"
+# Library Dependencies
+LIBRARY_DEPENDENCIES=""
```
Comment (adamlm) [inline install_dependencies.sh:7]: What is the difference between `DEPENDENCIES` and `LIBRARY_DEPENDENCIES`?

**Review by adamlm — CHANGES_REQUESTED** [review]

File: Config.cmake.in [hunk 20]
```diff
@@ -0,0 +1,11 @@
+@PACKAGE_INIT@
+
+include(CMakeFindDependencyMacro)
+
+find_dependency(${PROJ4_LIBRARIES} REQUIRED )
+find_dependency(units::units REQUIRED )
+find_dependency(GSL REQUIRED )
+
+
+check_required_components(@PROJECT_NAME@)
```
Comment (adamlm) [inline Config.cmake.in]: This is unnecessary and should be removed. It is from an older style of CMake.

File: Config.cmake.in [hunk 21]
```diff
@@ -0,0 +1,11 @@
+@PACKAGE_INIT@
```
Comment (adamlm) [inline Config.cmake.in:1]: This is unnecessary, and should be removed. The name should ideally be changed to remove the `.in`, but I believe we are limited to other CARMA CMake modules.

File: Config.cmake.in [hunk 22]
```diff
@@ -0,0 +1,11 @@
...
+
+include(CMakeFindDependencyMacro)
+
+find_dependency(${PROJ4_LIBRARIES} REQUIRED )
```
Comment (adamlm) [inline Config.cmake.in]: This should be `find_dependency(PROJ4 REQUIRED)`.

File: cmake/dependencies.cmake [hunk 23]
```diff
@@ -0,0 +1,49 @@
...
+  GITHUB_REPOSITORY nholthaus/units
+  GIT_TAG v2.3.3
+  SYSTEM ON
+  OPTIONS
+    "BUILD_TESTS FALSE"
+    "BUILD_DOCS FALSE"
+)
+#TODO: This is probably not a complete working fix but it allows that project install
+# target to run successfully. Needs tot be updated
+export( TARGETS units NAMESPACE units:: FILE unitsConfig.cmake)
```
Comment (adamlm) [inline cmake/dependencies.cmake]: This is unnecessary because the `units` library always creates install targets.

**Review by adamlm — CHANGES_REQUESTED** [review]

File: .devcontainer/devcontainer.json (hunk 1 above)
Comment (paulbourelly999) [reply .devcontainer/devcontainer.json]: Good call.

File: .vscode/tasks.json (hunk 4 above)
Comment (paulbourelly999) [reply .vscode/tasks.json:1]: It allows for the use of keyboard short cuts for building the library. It also kind of integrates the build process with our IDE (VS Code) to allow for developers to build the library without calling and scripts or commands directly.

File: .editorconfig (hunk 2 above)
Comment (paulbourelly999) [reply .editorconfig:1]: We can change to using what every tool for formatting if there is something better. I would suggest we stick with EditorConfig at least for this PR so we have something since we are already using it in other carma apt repos like carma-time-library. We can put in an issue to update this to a different tool and possibly include the configurations in a common place, like the carma-builds images. Just a suggestion

File: .vscode/launch.json (hunk 3 above)
Comment (paulbourelly999) [reply .vscode/launch.json]: Do we need any configuration file to setup the debugger for the CMake extension or does it work out of the box?

File: CMakeLists.txt (hunk 5 above)
Comment (paulbourelly999) [reply CMakeLists.txt]: Good catch

File: .vscode/launch.json (hunk 3 above)
Comment (adamlm) [reply .vscode/launch.json]: No, it should work out of the box.

File: CMakeLists.txt (hunk 6 above)
Comment (paulbourelly999) [reply CMakeLists.txt]: I think it may be more important for us to version these packages in a consistent way in this case than in the standard way. We need to rework the debian package release process anyway because I think there are still some outstanding things there as well. We are already versioning the carma-time-lib this way. If we do correct this or change this, we may want to do it for all the stol debian packages at once to be consistent.

File: CMakeLists.txt (hunk 9 above)
Comment (paulbourelly999) [reply CMakeLists.txt:20]: Hmmm i have seen a lot of projects do this including the units library you included. Could you maybe explain more about why it is important to avoid using ${PROJECT_NAME} in this way. Since this project is essentially just this library I do not see why this would be difficult to read or a concern.

File: CMakeLists.txt (hunk 13 above)
Comment (paulbourelly999) [reply CMakeLists.txt]: Good catch

File: CMakeLists.txt (hunk 12 above)
Comment (paulbourelly999) [reply CMakeLists.txt]: Is this true for CMake 3.10.2 or just newer versions?

File: CMakeLists.txt (hunk 15 above)
Comment (paulbourelly999) [reply CMakeLists.txt]: It looks like to me this is SOVERSION is used by several things in the CPack functionality.
https://github.com/usdot-fhwa-stol/carma-builds/blob/cd6e4ce3d443539ec4fd598d1c41225d79ecdcc4/cmake/Packing.cmake#L19

File: cmake/dependencies.cmake (hunk 23 above)
Comment (paulbourelly999) [reply cmake/dependencies.cmake]: When I remove this I get the following CMake error:
```
Make Error in CMakeLists.txt:
  export called with target "carma-geodetic" which requires target "units"
  that is not in any export set.
```

File: CMakeLists.txt (hunk 7 above)
Comment (paulbourelly999) [reply CMakeLists.txt:1]: Not sure what this is about? Are you saying the indents in the file should be 2 spaces note 4?

File: CMakeLists.txt (hunk 6 above)
Comment (adamlm) [reply CMakeLists.txt]: Yeah that's fine. I don't think there are any hard rules for versioning < 1.0.0.

File: CMakeLists.txt (hunk 9 above)
Comment (adamlm) [reply CMakeLists.txt:20]: Here is an excerpt from Craig Scott's _Professional CMake: A Practical Guide_, Section 4.6. Recommended Practices:

> Target names need not be related to the project name. It is common to see tutorials and examples use a variable for the project name and reuse that variable for the name of an executable [in our case, library] target:
>
> [example CMake snippet removed]
>
> This only works for the most basic of projects and encourages a number of bad habits. Consider the project name and executable [or library] name as being separate, even if initially they start out the same. Set the project name directly rather than via a variable, choose a target name according to what the target does rather than the project it is part of an assume the project will eventually need to define more than one target. This reinforces better habits which will be important when working on more complex multi-target projects [for example, ROS packages that define both libraries and executables].

File: CMakeLists.txt (hunk 12 above)
Comment (adamlm) [reply CMakeLists.txt]: Yes, this is included in CMake 3.10.

File: CMakeLists.txt (hunk 8 above)
Comment (paulbourelly999) [reply CMakeLists.txt]: Its required by linking to the test executable created by the https://github.com/usdot-fhwa-stol/carma-builds/blob/cd6e4ce3d443539ec4fd598d1c41225d79ecdcc4/cmake/Testing.cmake . To me it seems if it is really required there we should include it there. Not sure I understand this dependencies but have seen it before for creating test executable. Reading up on it to try and educate myself. Without this I get the following error though:
```
/usr/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/9/../../../x86_64-linux-gnu/libgtest.a(gtest-all.cc.o): undefined reference to symbol 'pthread_getspecific@@GLIBC_2.2.5'
/usr/bin/ld: /lib/x86_64-linux-gnu/libpthread.so.0: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
```

File: CMakeLists.txt (hunk 15 above)
Comment (adamlm) [reply CMakeLists.txt]: I see. In that case, I believe this should be `0` since the library is on version `0.0.1`. However, I'm not super familiar with soversion conventions and packaging.

Here's a link to a presentation Craig Scott made that has some relevant info (starting at slide 34): [link](https://github.com/CppCon/CppCon2019/blob/master/Presentations/deep_cmake_for_library_authors/deep_cmake_for_library_authors__craig_scott__cppcon_2019.pdf)

File: CMakeLists.txt (hunk 9 above)
Comment (paulbourelly999) [reply CMakeLists.txt:20]: I am ok with changing this to enforce good habits for defining CMakeLists.txt files. Want to make sure I understand the concern here. The concern is that defining targets with project name links the target to the project which should ideally be 2 independent concepts since more complex CMake files can have multiple targets. 

In this case for stol-apt repos my understanding is that ideally, these would always be single target projects  since any dependent targets should probably be independent repos pulled in via apt stol debian package repository rather than targets included in one larger monolithic CMake project right? Or do you see utility in having mutli-target repos here that get build in to debian package?

File: CMakeLists.txt (hunk 15 above)
Comment (paulbourelly999) [reply CMakeLists.txt]: I don't think you can set this to `0`. When I set it to `0` I get errors from the Packing.cmake that indicate it is not set at all 
```
CMake Error at /opt/carma/cmake/Packing.cmake:19 (message):
  Target property SOVERSION must be defined for shared library packages.
```

This kind of makes sense since I have never seen a shared library file/simlink by <library_name>.so.0 but I could be missing something here.

File: Config.cmake.in (hunk 21 above)
Comment (paulbourelly999) [reply Config.cmake.in:1]: I think this is all kind of based on this [tutorial](https://cmake.org/cmake/help/latest/guide/tutorial/Adding%20Export%20Configuration.html). It seems to match pretty closely with what is implemented in the carma-builds InstallConfigs.cmake . This tutorial seems to be up to date with the current CMake verison. Are we doing something else that makes this unnecessary?

File: build.sh (hunk 17 above)
Comment (paulbourelly999) [reply build.sh]: Yes calls the carma-builds build_script with parameters to generate the debian package. The carma-builds build_script is responsible for calling the appropriate cmake and cpack calls with the appropriate options set.

File: CMakeLists.txt (hunk 15 above)
Comment (adamlm) [reply CMakeLists.txt]: That appears to be a bug in the `Packing.cmake` file. The CMake documentation mentions [`get_target_property()`](https://cmake.org/cmake/help/latest/command/get_target_property.html) will set `<VAR>` to `<VAR>-NOTFOUND` or an empty string, depending on the context. The `if` state should be changed to something like this:

```cmake
if(NOT "${so_version}")
  # ...
endif()
```
I don't think I got the syntax quite right, but this shows the idea.

I don't see anywhere in the [CMake documentation](https://cmake.org/cmake/help/latest/prop_tgt/SOVERSION.html) preventing `SOVERSION` from being `0`, and I've seen examples of some projects having a `0` soversion.

File: build.sh (hunk 17 above)
Comment (paulbourelly999) [reply build.sh]: I guess we could replace this file by setting the vscode build task to call this directly but I think the script is useful for abstracting this from the developer and just allowing them to use either the vscode task or script with no parameters set.

File: CMakeLists.txt (hunk 7 above)
Comment (adamlm) [reply CMakeLists.txt:1]: Correct. That is CMake's styling convention.

File: install_dependencies.sh (hunk 18 above)
Comment (paulbourelly999) [reply install_dependencies.sh]: Good catch. Copied this from carma-time-lib where it probably also needs to be fixed. I would guess this is a copy paste mistake that is also present in carma-builds.

File: CMakeLists.txt (hunk 15 above)
Comment (paulbourelly999) [reply CMakeLists.txt]: Ok let me check this out and see if I can fix it.

File: CMakeLists.txt (hunk 9 above)
Comment (adamlm) [reply CMakeLists.txt:20]: Packaging is an orthogonal concept to project structure. A good project structure gives packagers the freedom to package things how they see fit.

Multi-target projects are essentially inherent to how we're developing. For example, a library is one target, and the unit tests are another. So in this case we already have multiple targets. I'm not sure if that answers your question.

File: CMakeLists.txt (hunk 8 above)
Comment (adamlm) [reply CMakeLists.txt]: I think this is a bug in `Testing.cmake`. It should be linking against `GTest::gtest_main` instead of `gtest`. Can you make that change locally and see if that resolves it?

File: Config.cmake.in (hunk 21 above)
Comment (adamlm) [reply Config.cmake.in:1]: Using the configuration we have now is a bit over-complicating things. `*Config.cmake.in` is when you have more elaborate dependency resolution. I have needed to use `*Config.cmake.in` only once or twice. Refactoring this makes the `*Config.cmake` install a simple `install(FILES)` call.

File: Config.cmake.in (hunk 20 above)
Comment (paulbourelly999) [reply Config.cmake.in]: Was reading the following documentation about this call [here](https://cmake.org/cmake/help/latest/module/CMakePackageConfigHelpers.html). Is this unnecessary in your mind because the to `configure_package_config_file ()` in carma-builds InstallingConfigs.cmake should generate this macros by default as long as we do not specific `NO_CHECK_REQUIRED_COMPONENTS_MACRO` option?

File: test/test_main.cpp [hunk 24]
```diff
@@ -0,0 +1,7 @@
+#include <gtest/gtest.h>
```
Comment (adamlm) [inline test/test_main.cpp]: We should not need to have this file. GoogleTest can provide a `main` function for us. See an example from `libcarma`:

```cmake
add_executable(libcarma_sae_j2735_tests
  test_angle.cpp
)

target_link_libraries(libcarma_sae_j2735_tests
  PRIVATE
    GTest::gtest_main
    libcarma::sae_j2735
)

include(GoogleTest)
gtest_discover_tests(libcarma_sae_j2735_tests)
```

File: CMakeLists.txt (hunk 7 above)
Comment (paulbourelly999) [reply CMakeLists.txt:1]: No sure this is terribly important but including it in the editorconfig just to be thorough.

**Review by adamlm — CHANGES_REQUESTED** [review]

File: .vscode/launch.json (hunk 3 above)
Comment (paulbourelly999) [reply .vscode/launch.json]: Maybe tomorrow you can show me how this works in the extension because I do not see anything in VSCode nor did I find anything only about setting up code debugger with the CMake extension.

File: CMakeLists.txt (hunk 8 above)
Comment (paulbourelly999) [reply CMakeLists.txt]: Works and added the following PR to fix it (https://github.com/usdot-fhwa-stol/carma-builds/pull/27/)

File: Config.cmake.in (hunk 21 above)
Comment (paulbourelly999) [reply Config.cmake.in:1]: It sounds like this is more of a preference on how this process should be setup rather than a case in which we are doing something wrong or incorrect. If this is the case I think having it consistent with our other CMake installation setups is better for right now. We can put an issue up for changing this process if there is some substantial benefit and pick it up if necessary.

File: CMakeLists.txt (hunk 15 above)
Comment (paulbourelly999) [reply CMakeLists.txt]: Addressed in PR https://github.com/usdot-fhwa-stol/carma-builds/pull/26

File: CMakeLists.txt (hunk 12 above)
Comment (paulbourelly999) [reply CMakeLists.txt]: After looking at the [tutorial](https://cmake.org/cmake/help/latest/guide/tutorial/Adding%20Export%20Configuration.html) I mentioned earlier it still seems that this convention is practiced. Since this accomplishes the same thing either way I think its probably better we stick with the convention for now to be consistent. For the sake of time and consistency I think making changes that are non-functional like this need more of a business justification but I still think its always a good idea to track any things we can potentially improve.

File: CMakeLists.txt (hunk 9 above)
Comment (paulbourelly999) [reply CMakeLists.txt:20]: If we want to address this I suggest we create a separate issue. I think most of the cmake files from the carma-builds image also depend on target which works out well in practice right now for the libraries that use  carma-builds since they just build a single project executable and test executable. I can see that this separation could be helpful if in the future other projects injest our libraries using the source code and CMake files although currently that is not really the intent of these libraries so doing work to enable this may be low priority for now.

File: install_dependencies.sh (hunk 19 above)
Comment (paulbourelly999) [reply install_dependencies.sh:7]: Not sure what Robs initial idea here was. The only thing that makes sense to me is one to capture direct dependencies and indirect dependencies (dependencies of dependencies). The only benefit I could see in separating them is maybe maintenance. 
It may be easier to ensure all dependencies included are necessary if there is a distinction between direct and indirect dependencies.
I am also willing to remove this since I can really think of strong justification for keeping this. I'm indifferent either way. I'll let you make the call.

File: test/test_main.cpp (hunk 24 above)
Comment (paulbourelly999) [reply test/test_main.cpp]: I'll try to incorporate this in the Testing.cmake in my current PR (https://github.com/usdot-fhwa-stol/carma-builds/pull/27). Seems appropriate there.

File: Config.cmake.in (hunk 21 above)
Comment (adamlm) [reply Config.cmake.in:1]: That makes sense. This is indeed a case where it's not wrong but could be improved. We'll punt this to separate task/issue.

File: CMakeLists.txt (hunk 9 above)
Comment (adamlm) [reply CMakeLists.txt:20]: I see what you mean. The carma-builds CMake modules all assume the targets use ${PROJECT_NAME} as their name, so we are forced to follow that in this case. Let's open a new issue in the [carma-builds](https://github.com/usdot-fhwa-stol/carma-builds) repo about improving their CMake modules.

File: test/test_main.cpp (hunk 24 above)
Comment (paulbourelly999) [reply test/test_main.cpp]: Push to https://github.com/usdot-fhwa-stol/carma-builds/pull/27 and removing file

File: Config.cmake.in (hunk 20 above)
Comment (adamlm) [reply Config.cmake.in]: No, it is unnecessary because modern CMake uses target-based configurations, and the targets should be responsible for checking that their dependencies are satisfied. Older CMake used instead of targets, so it was necessary to check that the variables were all defined properly. Here's the relevant excerpt from Craig Scott's book:

> In the past when all details about a package were provided through variables, it was customary to check whether all required variables were set at the end of the config file before returning. A macro called `check_required_components()` was defined for this purpose, but projects that provide imported targets should perform these checks themselves... This makes the `check_required_componetns()` macro largely redundant.

Comment (paulbourelly999) [reply Config.cmake.in]: Personally I think this sounds more like a preference Craig Scott has in setting of CMake configurations to move responsibility for checking required components to the consuming project since the CMake official documentation still suggests we use this macros for CMake configs. I am not opposed to using a better approach as long as the benefit of the change is worth the effort involved. In this case, I think if we want to change our CMake library install configurations to match closer to Craig Scotts approach I suggest we do this in a separate PR and issue so we can prioritize this work against the other improvements we need to make and so that we can implement this change more consistently across projects

File: cmake/dependencies.cmake (hunk 23 above)
Comment (adamlm) [reply cmake/dependencies.cmake]: This is a problem with the carma-builds `InstallingConfig.cmake` file, which makes a call to `export(TARGETS)`. I believe this requires all targets in the export set to also have their import targets available immediately. I could be wrong here though. In any case, I don't think the `export(TARGETS)` call is need here nor in `InstallingConfig.cmake`. We should open a new issues in the carma-builds repo about this.

File: install_dependencies.sh (hunk 19 above)
Comment (adamlm) [reply install_dependencies.sh:7]: Let's keep it for now. I'm not knowledgeable enough right now to form an opinion on it.

File: CMakeLists.txt (hunk 12 above)
Comment (adamlm) [reply CMakeLists.txt]: That's fine. We can keep it for now and make an issue elsewhere for an eventual migration. A lot of projects still do it this way, so no big deal.

**Review by adamlm — APPROVED** [review]
